featurebyte/utils/async_helper.py

```diff
@@ -0,0 +1,55 @@
+"""
+Helper functions for asyncio
+"""
+
+from __future__ import annotations
+
+from typing import Any, Coroutine, List
+
+import asyncio
+from asyncio import Future
+
+
+def asyncio_gather(*coros: Coroutine[Any, Any, Any], max_concurrency: int = 0) -> Future[List[Any]]:
+    """
+    Run coroutines with a optional concurrency limit
+
+    Parameters
+    ----------
+    coros: Coroutine[Any, Any, Any]
+        Coroutines to run
+    max_concurrency: int
+        Maximum number of coroutines to run concurrently
+
+    Returns
+    -------
+    Future[List[Any]]
+    """
+
+    if max_concurrency < 1:
+        return asyncio.gather(*coros)
+
+    failed = False
+    tasks = []
+    semaphore = asyncio.Semaphore(max_concurrency)
+
+    # Run coroutines with a semaphore
+    async def _coro(coro: Coroutine[Any, Any, Any]) -> Any:
+        async with semaphore:
+            nonlocal failed, tasks
+            if failed:  # pylint: disable=used-before-assignment
+                # Close unawaited coroutines on failure
+                coro.close()
+                return
+            task = asyncio.create_task(coro)
+            tasks.append(task)
+            try:
+                return await task
+            except Exception:
+                failed = True
+                # Cancel all tasks on failure
+                for task in tasks:
+                    task.cancel()
```

Comment:
This loop over tasks is done within each coroutine, is that intended? It seems each task can potentially be cancelled more than once.

Reply:
This is intended as the caller of the `asyncio_gather` function has no access to the tasks created by the `_coro` wrapper functions. It seems that calling cancel on the same task multiple times does raise any error but I can add a flag to avoid the multiple calls